### PR TITLE
8343394: Make MemorySessionImpl.state a stable field

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch makes `MemorySessionImpl.state` a `Stable` field so that liveness check of non-closeable scopes such as the global scope can be elided.

Currently, the `state` field is overloaded with 2 responsibilities, to act as a communication device between `close` and `checkValidState`, as well as a communication device between `close`, `acquire`, and `release`. This patch separates those concerns into `state` and `acquireCount`, allowing `state` to be marked as `@Stable`.

With the patch, in `MemorySegmentGetUnsafe`, `panama` is able to be on par with `unsafe`:

    Benchmark                      Mode  Cnt  Score   Error  Units
    MemorySegmentGetUnsafe.panama  avgt   30  0.340 ± 0.008  ns/op
    MemorySegmentGetUnsafe.unsafe  avgt   30  0.332 ± 0.004  ns/op

For reference this is the results without this patch:

    Benchmark                      Mode  Cnt  Score   Error  Units
    MemorySegmentGetUnsafe.panama  avgt   30  0.420 ± 0.019  ns/op
    MemorySegmentGetUnsafe.unsafe  avgt   30  0.329 ± 0.003  ns/op

Please kindly review, thanks very much.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8343394](https://bugs.openjdk.org/browse/JDK-8343394): Make MemorySessionImpl.state a stable field (**Enhancement** - P4)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**)

### Contributors
 * Maurizio Cimadamore `<mcimadamore@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21810/head:pull/21810` \
`$ git checkout pull/21810`

Update a local copy of the PR: \
`$ git checkout pull/21810` \
`$ git pull https://git.openjdk.org/jdk.git pull/21810/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21810`

View PR using the GUI difftool: \
`$ git pr show -t 21810`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21810.diff">https://git.openjdk.org/jdk/pull/21810.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21810#issuecomment-2450238165)
</details>

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/SharedSession.java
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/SharedSession.java:90]: I like it that we still only have one CAS here (as only one thread can set CLOSED_ACQUIRE_COUNT). So shared arena close doesn't need more work. You might want to check the MemorySessionClose bench, just in case.

File: src/java.base/share/classes/jdk/internal/foreign/MemorySessionImpl.java
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/MemorySessionImpl.java:61]: Nice trick! This effectively relegates 0 (open) to be the default value of the stable state field. That will either go to -1 (closed) or 1 (non-closeable) and be treated as constant thereafter. Perhaps worth adding a comment on the rationale behind this.

**Comment by merykitty** [comment]
The result of the new benchmark by @mcimadamore , the patch offers an improvement in the `segment_loop_all` and `segment_loop_asUnchecked` cases. For some reasons, the compiler really dislikes the unsafe loop.

                                                           Before          After
    Benchmark                                Mode  Cnt  Score   Error  Score   Error  Units
    LoopOverRandom.segment_loop              avgt   30  2.218 ± 0.007  2.261 ± 0.096  us/op
    LoopOverRandom.segment_loop_all          avgt   30  1.958 ± 0.007  1.917 ± 0.005  us/op
    LoopOverRandom.segment_loop_asUnchecked  avgt   30  1.834 ± 0.005  1.747 ± 0.004  us/op
    LoopOverRandom.unsafe_loop               avgt   30  1.825 ± 0.026  1.795 ± 0.004  us/op

**Comment by mcimadamore** [comment]
> For some reasons, the compiler really dislikes the unsafe loop.

On my slower machine, the difference between the variants is bigger:

```
Benchmark                                Mode  Cnt  Score   Error  Units
LoopOverRandom.segment_loop              avgt   15  3.727 ± 0.063  us/op
LoopOverRandom.segment_loop_all          avgt   15  2.813 ± 0.100  us/op
LoopOverRandom.segment_loop_asUnchecked  avgt   15  2.674 ± 0.051  us/op
LoopOverRandom.unsafe_loop               avgt   15  2.785 ± 0.068  us/op
```

As you observed, `unsafe_loop` gives odd numbers.

File: src/java.base/share/classes/jdk/internal/foreign/ConfinedSession.java
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/ConfinedSession.java:44]: Since you are here, maybe you want to use `MHUtil::findVarHandle` also for the state var handle (for uniformity)

**Review by mcimadamore — APPROVED** [review]
Looks really good! Maybe you want to wait for @JornVernee review as well (this area is tricky - better to have another pair of eyes looking at it)

**Comment by merykitty** [comment]
Fixed the benchmark to get rid of `ValueLayout.scale` which distorts the result, now the results are:

                                                           Before          After
    Benchmark                                Mode  Cnt  Score   Error  Score   Error  Units
    LoopOverRandom.segment_loop              avgt   30  2.203 ± 0.008  2.258 ± 0.073  us/op
    LoopOverRandom.segment_loop_all          avgt   30  1.948 ± 0.006  1.771 ± 0.005  us/op
    LoopOverRandom.segment_loop_asUnchecked  avgt   30  1.847 ± 0.014  1.746 ± 0.004  us/op
    LoopOverRandom.unsafe_loop               avgt   30  1.703 ± 0.005  1.716 ± 0.008  us/op

File: src/java.base/share/classes/jdk/internal/foreign/ConfinedSession.java
Comment (merykitty) [reply src/java.base/share/classes/jdk/internal/foreign/ConfinedSession.java:44]: I'm a little confused, I use `MHUtil::findVarHandle` for `acquireCount`, there is no need for a `VarHandle` for `state`.

Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/ConfinedSession.java:44]: Apologies, I got confused.

**Comment by mcimadamore** [comment]
Actually, @JornVernee pointed out (great observation!) that this PR creates a new asymmetry between liveness and acquired state -- it would be possible to see an acquire failure (e.g. when passing a segment to a native call) while the segment scope still looks alive. We need to think about this some more.

**Comment by merykitty** [comment]
Yes you are right, that is a great observation! My solution is that since we are throwing an exception anyway, we will spin on the `state` field to wait for it to change before throwing the exception.

File: src/java.base/share/classes/jdk/internal/foreign/SharedSession.java
Comment (minborg) [inline src/java.base/share/classes/jdk/internal/foreign/SharedSession.java:90]: Why are we using opaque semantics here and not volatile?

Comment (merykitty) [reply src/java.base/share/classes/jdk/internal/foreign/SharedSession.java:90]: Because this variable is not used as a synchronization device, so a stronger order is unnecessary. We still need the store to be observed by other threads when they fail an `acquire` or a `close` so opaque is appropriate here.

Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/SharedSession.java:90]: I'm dubious about this. Typically these modes work in a symmetric way - if you `setOpaque` you need to `getOpaque`. From [this](https://gee.cs.oswego.edu/dl/html/j9mm.html):

> Opaque mode, obtained using VarHandle getOpaque and setOpaque, adds constraints over Plain mode that provide minimal awareness of a variable that is subject to interthread access, **when all accesses use Opaque (or stronger) mode**

Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/SharedSession.java:90]: > I'm dubious about this. 

NVM, I see it now -`sharedSessionAlreadyAcquired` uses a `getVolatile, so we're in the clear.

File: src/java.base/share/classes/jdk/internal/foreign/MemorySessionImpl.java
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/MemorySessionImpl.java:62]:
```suggestion
     * is OPEN -> CLOSED. As a result, the states CLOSED and NONCLOSEABLE are stable. This allows us
```

File: src/java.base/share/classes/jdk/internal/foreign/SharedSession.java
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/SharedSession.java:101]: It would be good to add a test that tries to trigger this use case. e.g. have 2 threads which share a shared arena. One will try to close the arena, while the other will try to pass a segment allocated from the arena to a downcall.

File: src/java.base/share/classes/jdk/internal/foreign/SharedSession.java
Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/SharedSession.java:90]: Doesn't using `setOpaque` mean that another thread may see the update to `state` before the update to `acquireCount`? i.e. the scope of a memory segment may appear closed, but the segment would still be passable to a downcall?

File: test/micro/org/openjdk/bench/java/lang/foreign/LoopOverRandom.java
Comment (JornVernee) [inline test/micro/org/openjdk/bench/java/lang/foreign/LoopOverRandom.java:65]: Please use a fixed seed here, to avoid instability due to data being different.

File: src/java.base/share/classes/jdk/internal/foreign/SharedSession.java
Comment (ExE-Boss) [reply src/java.base/share/classes/jdk/internal/foreign/SharedSession.java:90]: `acquireCount` is always accessed using `volatile` semantics by this `SharedSession`.

Comment (minborg) [reply src/java.base/share/classes/jdk/internal/foreign/SharedSession.java:90]: The close operation is not performance-critical so I think using volatile semantics here would have little or no negative performance impact. From a maintenance and readability point of view (in my subjective opinion) volatile is easier to understand.

Comment (merykitty) [reply src/java.base/share/classes/jdk/internal/foreign/SharedSession.java]: @JornVernee The acquire property of the `acquireCount` compare exchange prevents the `state` store from floating above it. But following @minborg suggestion I changed it to a volatile store.

**Comment by merykitty** [comment]
@JornVernee Thanks for your reviews, I have addressed them in the last commit.

File: test/jdk/java/foreign/TestMemorySession.java
Comment (JornVernee) [inline test/jdk/java/foreign/TestMemorySession.java:374]: I think the right primitive here is a `CyclicBarrier` with 2 parties. Each thread will then wait for the other before continuing on the next iteration.

Comment (merykitty) [reply test/jdk/java/foreign/TestMemorySession.java:374]: I'm afraid that it would be too expensive compared to a spinlock, and the scheduler might not wake up the threads soon enough to make a meaningful race, greatly reducing the effectiveness of the test.

Comment (JornVernee) [reply test/jdk/java/foreign/TestMemorySession.java:374]: Okay, fair enough.

**Review by JornVernee — APPROVED** [review]

**Comment by minborg** [comment]
In `SharedSession`, would it not be necessary to override `isClosable()` so it would be using `volatile` semantics to read the `closed` field rather than relying on plain reads?

File: test/jdk/java/foreign/TestMemorySession.java
Comment (mcimadamore) [reply test/jdk/java/foreign/TestMemorySession.java:374]: Using two blocking queues might also be possible:
* at the start of each iteration a thread does a `get` on its own queue
* at the end of each iteration, a thread does a put on the other thread's queue
* we call put on the first thread's queue, to get things moving

This should create the desired "ping-pong" effect.

Comment (merykitty) [reply test/jdk/java/foreign/TestMemorySession.java:374]: There are many ways to achieve this lock-step behaviour. In this case, the operation is extremely fast and there are only 2 threads waiting for each other so I believe a spinlock is the most appropriate. The major benefit of spinning is that we can be more confident that the threads will start each step at the approximately same time, increasing the chance to have a meaningful race.

**Comment by merykitty** [comment]
Thanks a lot for your reviews and suggestions, I will merge this patch.

**Comment by minborg** [comment]
Thanks for your contribution @merykitty ! I am already curious if you can find additional improvements!

**Comment by merykitty** [comment]
@mcimadamore Could you please re-approve this patch, I still have not got used to the stricter integration check :)

**Review by mcimadamore — APPROVED** [review]

**Pull request closed without merging**
